connectors/bitbucket-server/src/main/java/com/vmware/connectors/bitbucket/server/BitbucketServerController.java

```diff
@@ -0,0 +1,484 @@
+                                                @RequestHeader(AUTH_HEADER) final String authHeader,
+                                                @RequestHeader(BASE_URL_HEADER) final String baseUrl) {
+
+        logger.debug("Decline ACTION for bitbucket server pull request: {}, baseURL: {}",
+                pullRequest,
+                baseUrl);
+
+        return performBitbucketServerAction(baseUrl, authHeader, pullRequest, BitbucketServerAction.DECLINE);
+    }
+
+    @PostMapping(
+            path = "/api/v1/{projectKey}/{repositorySlug}/{pullRequestId}/comments",
+            consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE
+    )
+    public Single<ResponseEntity<String>> comments(final BitbucketServerPullRequest pullRequest,
+                                                  @RequestParam(COMMENT_PARAM_KEY) final String comment,
+                                                  @RequestHeader(AUTH_HEADER) final String authHeader,
+                                                  @RequestHeader(BASE_URL_HEADER) final String baseUrl) {
+
+        logger.info("Comment ACTION for bitbucket server pull request: {}, baseURL: {}",
+                pullRequest,
+                baseUrl);
+
+        final HttpHeaders headers = buildHeaders(authHeader);
+        headers.add(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE);
+
+        final URI uri = UriComponentsBuilder.fromHttpUrl(baseUrl + "/rest/api/1.0/projects/{PROJECT_KEY}/repos/{REPOSITORY_SLUG}/pull-requests/{PULL_REQUEST_ID}/comments")
+                .buildAndExpand(
+                        ImmutableMap.of(
+                                PROJECT_KEY, pullRequest.getProjectKey(),
+                                REPOSITORY_SLUG, pullRequest.getRepositorySlug(),
+                                PULL_REQUEST_ID, pullRequest.getPullRequestId()
+                        )
+                ).toUri();
+
+        final BitbucketServerComment bitBucketServerComment = new BitbucketServerComment(comment);
+        final ListenableFuture<ResponseEntity<String>> response = this.rest.exchange(
+                uri,
+                HttpMethod.POST,
+                new HttpEntity<>(bitBucketServerComment, headers),
+                String.class);
+        return Async.toSingle(response);
+    }
+
+    @GetMapping("/test-auth")
+    @SuppressWarnings("PMD.JUnit4TestShouldUseTestAnnotation") // PMD assumes the method as Junit Test if the API url has 'test'.
+    public Single<ResponseEntity<Void>> testAuth(@RequestHeader(AUTH_HEADER) final String authHeader,
+                                                 @RequestHeader(BASE_URL_HEADER) final String baseUrl) {
+        final HttpHeaders headers = new HttpHeaders();
+        headers.add(AUTHORIZATION, authHeader);
+
+        final ListenableFuture<ResponseEntity<JsonDocument>> response = this.rest.exchange(
+                baseUrl,
```

Comment:
I suspect this either isn't going to work, or it can be simpler.  The client doesn't actually use this endpoint yet, so you'll have to test it with postman or curl.

What should happen is:

```
curl -H "Authorization: Bearer <vidmToken>" -H "X-Bitbucket-Servicer-Authorization: Basic <base64EncodeUserColonPassword>" 'http://localhost:port/test-auth'
```

Should return 200 if the base64EncodedUserColonPassword is correct, or 401 if not.

So if my user/pass was: user1/password1, then my x-auth header would be: "Basic dXNlcjE6cGFzc3dvcmQx" (encoding obtained by: `echo -n "user1:password1 | base64`) and I should expect a 200.

If I try that with a different password "Basic dXNlcjE6aW52YWxpZA==" (encoding obtained by: `echo -n "user1:invalid" | base64`), then I should expect a 401.

Since you are only requesting the baseUrl (instead of looking up a PR or something), you don't need the 404To200 conversion (in the Jira connector, that was to handle the situation where the test ticket id XYZ-123 didn't exist).

Reply:
I confirmed that curling the base url does work (401 if creds are bad, 302 moved if creds are good, 302 moved if creds are blank).

I don't think the client will ever send in blank credentials "X-...-Authorization: Basic", so I'm ok w/ that returning 302.

We might want to map the 302 to 200, but it's probably fine as a 302 (we can revisit this when the client actually implements this).

Reply:
If the base url is not correct, then it will throw 404. In this case, we are mapping it 200.

Reply:
Why should a misconfiguration report a 200 (the user's credentials are correct) when the client calls into test-auth?

Reply:
Don't worry about this one, I created APF-776 to followup on this (for both jira and bitbucket).